### ISSUE TITLE
Resolve object lazily

### ISSUE THREAD

**Issue by AnisovAleksey** [Swinject/Swinject#303]
Hello!

In some cases we need to create wrapper like a ViewFactory or something like it. In general it is lazy initialization, so would be nice if Swinject will support lazy initialization.
It can be implemented as struct that contains Closure to resolve object. example of implementation below:

```
// Object will be common if someone have a strong reference
public struct Lazy<Service> {
    private let initClosure: () -> Service?
    private var service: Service?
    
    internal init(initClosure: () -> Service?) {
        self.initClosure = initClosure
    }
    
    public func resolve() -> Service? {
        if service == nil {
            service = initClosure()
        }
        return service
    }
}
```

What you think about it?

**Comment by jakubvano** [comment]
Hello @threat70, thanks for the suggestion! 👍 

I understand your reasoning, and can see something like this being added to the Swinject. Do you have some suggestions how an API might look like / some references to such feature in different DI frameworks?

I see a couple of issues (i.e. how would `ObjectScope.graph` work with lazy instantiation?), but i think it may be doable.

**Comment by threat70** [comment]
Hello @jakubvano.

I think that "lazy" should be a new method of registration, I mean that use something like `container.registerLazy(CustomView.self) { .... }` that will return `Lazy<Service>`, object scope will affect only `Lazy<Service>` object. `Lazy<Service>` store the last resolved instance of `Service` and provide it, so if you have the same `Lazy<Service>` object in other place, you will have the same `Service` at this places.

Also for another case can be implemented object like `Provider<Service>`, that will resolve `Service` for each call. 

I took it from Dagger 2+. ref: https://google.github.io/dagger/api/2.10/dagger/Lazy.html

Thanks

**Comment by J7mbo** [comment]
How about just providing an optional `lazy` argument to the current api?

**Comment by AnisovAleksey** [comment]
@J7mbo what do you actually mean? could you provide sample code?

**Comment by J7mbo** [comment]
@AnisovAleksey I tried something the other day, whereby I inject a closure returning the object instead. 

```
container.register(MyObj.self) { in
    return { MyObj() 
}
```

Then in your object that requires `MyObj`, you type hint for the closure:

```
class MyClass
{
     public var myObj: (() -> (MyObj))?
}
```

So when `MyClass` is initialised (and resolved from the container), only the closure containing `MyObj` is. You don't actually instantiate the object until, somewhere in your code, you call `self.myObj()`.

Assuming the above works (as I didn't properly test it), I would propose a `lazy` argument to `register()`, which basically just wraps it in a closure instead. You'd still have to ask for a closure in your `MyClass` though, I'm not sure how we would fix that one unless we generated a subclass of MyObj, sort of a proxy object, which when called returns the actual Object...

Just brain dumping here :)

**Comment by AnisovAleksey** [comment]
@J7mbo I'm not sure, that registration of closure will work.
This feature can implemented as Extensions to resolver that will wrap call of `resolve()` function to one of two types: **Lazy** or **Provider**.
**Provider** - always create new one
**Lazy** - create object only one time, and then use it

sample code below: 
```
public func resolveLazy<Service>(_ serviceType: Service.Type, name: String?) -> Lazy<Service?> {
    return Lazy<Service?> {
        self.resolve(serviceType, name: name)
    }
}

public func resolveProvider<Service>(_ serviceType: Service.Type, name: String?) -> Provider<Service?> {
    return Provider<Service?> {
        self.resolve(serviceType, name: name)
    }
}
```

Lazy:
```
public class Lazy<Service> {
    private let resolveClosure: () -> Service?
    private var service: Service?
    
    internal init(resolveClosure: @escaping () -> Service?) {
        self.resolveClosure = resolveClosure
    }
    
    public func resolve() -> Service? {
        if service == nil {
            service = resolveClosure()
        }
        return service
    }
}
```

Provider:
```
public class Provider<Service> {
    private let resolveClosure: () -> Service?
    
    internal init(resolveClosure: @escaping () -> Service?) {
        self.resolveClosure = resolveClosure
    }
    
    public func resolve() -> Service? {
        return resolveClosure()
    }
}
```

Current problem is that it breaks _SwinjectAutoregistration_ (not actually tested, but I think that problems with type inference will appears). So maybe should be one `resolve()` method that will check for Lazy and Provider type. 
Do this on registering isn't right because we shouldn't think about Lazy while registrate new object. Lazy - is resolve type of **any** dependency.

**Comment by J7mbo** [comment]
@AnisovAleksey I've just tried and I am absolutely 100% sure that you simply say that the dependency in a closure, and it is only instantiated when you call it yourself. It's very simple. The registration of closure does indeed work, try it!

**Comment by jakubvano** [comment]
@AnisovAleksey I agree, we should distinguish between `Provider` and `Lazy` wrappers - the `Provider` functionality could be achieved by simple closures, as @J7mbo suggested, but it is cleaner to put these 2 behaviours on the same level, i.e. creating wrapper types for each.

I've been thinking about this, and IMO the ideal API would be no registration API 😄 - you would only need to register the type itself, and registrations for the `Lazy` and `Provider` would happen automatically -  this could be achieved by implementing general support for Behaviours - #322, and then adding behaviours for lazy / provider generation. 
One thing I'm not sure about though is how would the 'automatic' registration infer the object scope of `Lazy` / `Provider`.

**Comment by AnisovAleksey** [comment]
@jakubvano registration Lazy and Provider for any type will cause performance issue, because on app startup everyone has a lot of registrations and these will be longer by 3 times. registration of additional types for everything will be fine If you know how to resolve performance issue.

Object scope should be the same as for base object (that wrapped in Lazy), but still have questions about Graph scope... I think it requires more investigation on other libraries

**Comment by J7mbo** [comment]
@AnisovAleksey On this point, I have seen other libraries for other languages _cache_ the injections definitions, or the object graph. Maybe this is something we can do? 

A side-point, if we ever get good enough reflection support in swift, we can have **true** auto-injection, and by that I mean automatically reflecting on all the initialisers of objects that will be instantiated and wiring up everything without you even having to register them, as long as they are concrete implementations.

**Comment by AnisovAleksey** [comment]
@J7mbo Ahh... this increases the complexity of implementation :(   But yes, I think this is the only way out, we should cache Graph where we resolve Lazy object.

**Comment by jakubvano** [comment]
@AnisovAleksey You are right, if we generated `Lazy` registrations for each service, it might significantly impact the performance. Will look into possibilities..

In the meantime, I started implementing basic support for lazy injection and graph caching (#331 ). Would love your input 😉

**Comment by jakubvano** [comment]
Implemented by #331
